projects/clr/hipamd/include/hip/amd_detail/amd_device_functions.h

```diff
@@ -751,60 +759,93 @@ __device__ inline __attribute__((convergent)) int __syncthreads_or(int predicate
   STATE_ID    29:27   State ID (graphics only, not compute).
   ME_ID       31:30   Micro-engine ID.
 
+  HW_ID Register bit structure for RDNA4
+  WAVE_ID     4:0     Wave id within the SIMD.
+  SIMD_ID     9:8     SIMD_ID within the WGP: [0] = CU within WGP, [1] = SIMD within CU.
+  WGP_ID      13:10   Physical WGP ID.
+  SA_ID       16      Shader Array ID.
+
   XCC_ID Register bit structure for 942/950
   XCC_ID      3:0     XCC the wave is assigned to.
+
+  Encoding of parameter bitmask
+  HW_ID        5:0     HW_ID
+  OFFSET       10:6    Range: 0..31
+  SIZE         15:11   Range: 1..32
  */
 
-#if (defined(__GFX10__) || defined(__GFX11__))
-#define HW_ID 23
-#else
-#define HW_ID 4
-#endif
+#define GETREG_IMMED(SZ, OFF, REG) (((SZ) << 11) | ((OFF) << 6) | (REG))
+
+__device__ inline unsigned __smid_gfx1x() {
+  constexpr unsigned int kSEIdSize = 3, kSEIdOffset = 18, kHwId = 23, kWgpIdSize = 4,
+                         kWgpIdOffset = 10, kSAIdSize = 1, kSAIdOffset = 16;
+  unsigned se_id = __builtin_amdgcn_s_getreg(GETREG_IMMED(kSEIdSize - 1, kSEIdOffset, kHwId)),
+           wgp_id = __builtin_amdgcn_s_getreg(GETREG_IMMED(kWgpIdSize - 1, kWgpIdOffset, kHwId)),
+           sa_id = __builtin_amdgcn_s_getreg(GETREG_IMMED(kSAIdSize - 1, kSAIdOffset, kHwId)),
+           cu_id = 0;
 
-#if (defined(__GFX10__) || defined(__GFX11__))
-#define HW_ID_WGP_ID_SIZE 4
-#define HW_ID_WGP_ID_OFFSET 10
 #if (defined(__AMDGCN_CUMODE__))
```

Comment:
This won't work with SPIR-V, we'll need to add a predicate for it; at least a comment would help, but a warning printed out at compile time might be nicer (failing to compile would be harsh).